src/arch/src/aarch64/gic/regs/mod.rs

```diff
@@ -2,17 +2,132 @@ mod dist_regs;
 mod icc_regs;
 mod redist_regs;
 
-use crate::aarch64::gic::Result;
-
+use crate::aarch64::gic::{Error, Result};
+use kvm_bindings::kvm_device_attr;
 use kvm_ioctls::DeviceFd;
+use std::iter::StepBy;
+use std::ops::Range;
 use versionize::{VersionMap, Versionize, VersionizeResult};
 use versionize_derive::Versionize;
 
+pub trait MmappedReg {
+    fn range(&self) -> Range<u64>;
+
+    fn iter<T>(&self) -> StepBy<Range<u64>>
+    where
+        Self: Sized,
+    {
+        self.range().step_by(std::mem::size_of::<T>())
+    }
+}
+
+pub trait VgicRegEngine {
+    type Reg: MmappedReg;
+    type RegChunk: Default + Clone;
+
+    fn group() -> u32;
+
+    fn mpidr_mask() -> u64;
+
+    fn kvm_device_attr(offset: u64, val: &mut Self::RegChunk, mpidr: u64) -> kvm_device_attr {
+        kvm_device_attr {
+            group: Self::group(),
+            attr: (mpidr & Self::mpidr_mask()) | offset,
+            addr: val as *mut Self::RegChunk as u64,
+            flags: 0,
+        }
+    }
+
+    fn get_reg_data(fd: &DeviceFd, reg: &Self::Reg, mpidr: u64) -> Result<Vec<Self::RegChunk>>
```

Comment:
nit:
since `get_reg_data()` is not really used on its own, I would inline it in `get_regs_data()` and avoid an extra dynamic dispatch function call.

Reply:
Good idea. For the moment I marked it as `inline`. I wouldn't remove the method since we might need to use it in the future. It offers extra flexibility.